File: app/src/main/java/com/fortitude/shamsulkarim/ieltsfordory/data/repository/FirebaseAudioRepository.java
package com.fortitude.shamsulkarim.ieltsfordory.data.repository;

import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.io.File;
import java.io.IOException;

public class FirebaseAudioRepository implements AudioRepository {
    private final FirebaseStorage storage;

    public FirebaseAudioRepository() {
        this.storage = FirebaseStorage.getInstance();
    }

    @Override
    public void downloadAudio(String wordName, Callback callback) {
        String lowerCaseWordName = wordName.toLowerCase();
        StorageReference gsReference = storage
                .getReferenceFromUrl(
                        "gs://fir-userauthentication-f751c.appspot.com/audio/" + lowerCaseWordName + ".mp3");

        File localFile;
        try {
            localFile = File.createTempFile("Audio", "mp3");
        } catch (IOException e) {
            callback.onError(e);
            return;
        }

        gsReference.getFile(localFile).addOnSuccessListener(taskSnapshot -> {
            callback.onAudioReady(localFile);
        }).addOnFailureListener(e -> {
            callback.onError(e);
        });
    }
}

File: app/src/main/java/com/fortitude/shamsulkarim/ieltsfordory/data/media/firebase/FirebaseAudioRepository.kt
package com.fortitude.shamsulkarim.ieltsfordory.data.media.firebase

import com.fortitude.shamsulkarim.ieltsfordory.domain.media.AudioRepository
import com.fortitude.shamsulkarim.ieltsfordory.domain.media.model.AudioData
import com.google.firebase.storage.FirebaseStorage
import java.io.File
import java.io.IOException

class FirebaseAudioRepository : AudioRepository {
    private val storage: FirebaseStorage = FirebaseStorage.getInstance()

    override fun downloadAudio(wordName: String, callback: AudioRepository.Callback) {
        val lowerCaseWordName = wordName.lowercase()
        val ref = storage.getReferenceFromUrl(
            "gs://fir-userauthentication-f751c.appspot.com/audio/$lowerCaseWordName.mp3"
        )

        val localFile: File = try {
            File.createTempFile("Audio", "mp3")
        } catch (e: IOException) {
            callback.onError(e)
            return
        }

        ref.getFile(localFile).addOnSuccessListener {
            callback.onSuccess(AudioData(localFile.absolutePath))
        }.addOnFailureListener { e ->
            callback.onError(e)
        }
    }
}

